File: src/Cxx/Utilities/ResetCameraOrientation.cxx
#include <vtkActor.h>
#include <vtkCamera.h>
#include <vtkConeSource.h>
#include <vtkInteractorStyleTrackballCamera.h>
#include <vtkNamedColors.h>
#include <vtkNew.h>
#include <vtkPolyDataMapper.h>
#include <vtkProperty.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderer.h>

#include <chrono>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

namespace {

//! The positional information relating to the camera.
struct Orientation
{
  void ToVector(const double* from, std::vector<double>& to)
  {
    for (size_t i = 0; i < to.size(); ++i)
    {
      to[i] = from[i];
    }
  };

  std::vector<double> position{0.0, 0.0, 0.0};
  std::vector<double> focalPoint{0.0, 0.0, 0.0};
  std::vector<double> viewUp{0.0, 0.0, 0.0};
  double distance = 0;
  std::vector<double> clipplingRange{0.0, 0.0};
  std::vector<double> orientation{0.0, 0.0, 0.0};
};

//! Produce a comma-separated string of numbers from a vector.
/*
 *
 * @param v - the vector of doubles.
 * @param precision - number of decimal places.
 * @param width - the width.
 * @return A comma-separated string.
 */
std::string CommaSeparatedList(std::vector<double> const& v,
                               const int precision = 6, const int width = 10);

//! Get the camera orientation.
/*
 * @param ren - the renderer.
 * @return The orientation parameters.
 */
Orientation GetOrientation(vtkRenderer* ren);

//! Set the camera orientation.
/*
 *
 * @param ren - the renderer.
 * @param p - The orientation parameters.
 * @return
 */
void SetOrientation(vtkRenderer* ren, Orientation const& p);

} // namespace

int main(int, char*[])
{
  vtkNew<vtkNamedColors> colors;

  vtkNew<vtkConeSource> cone;
  cone->SetHeight(3.0);
  cone->SetRadius(1.0);
  cone->SetResolution(10);

  vtkNew<vtkPolyDataMapper> coneMapper;
  coneMapper->SetInputConnection(cone->GetOutputPort());

  vtkNew<vtkActor> coneActor;
  coneActor->SetMapper(coneMapper);
  coneActor->GetProperty()->SetColor(colors->GetColor3d("Bisque").GetData());

  vtkNew<vtkRenderer> ren;
  ren->AddActor(coneActor);
  ren->SetBackground(colors->GetColor3d("MidnightBlue").GetData());

  vtkNew<vtkRenderWindow> renWin;
  renWin->AddRenderer(ren);
  renWin->SetSize(600, 600);
  renWin->SetWindowName("ResetCameraOrientation");

  vtkNew<vtkRenderWindowInteractor> iren;
  iren->SetRenderWindow(renWin);

  vtkNew<vtkInteractorStyleTrackballCamera> style;
  iren->SetInteractorStyle(style);

  vtkCamera* camera = ren->GetActiveCamera();
  camera->SetRoll(15);
  camera->Elevation(-15);
  camera->Azimuth(30);
  ren->ResetCamera();

  renWin->Render();
  Orientation originalOrient = GetOrientation(ren);
  std::cout << std::setw(23) << "Original orientation:"
            << CommaSeparatedList(originalOrient.orientation) << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(1));

  camera->SetPosition(-3.568189, 5.220048, 2.352639);
  camera->SetFocalPoint(-0.399044, -0.282865, 0.131438);
  camera->SetViewUp(0.623411, 0.573532, -0.531431);
  camera->SetDistance(6.727500);
  camera->SetClippingRange(3.001430, 11.434082);
  // No need to use ren->ResetCamera() as we have all the parameters.
  renWin->Render();
  Orientation newOrient = GetOrientation(ren);
  std::cout << std::setw(23)
            << "New orientation:" << CommaSeparatedList(newOrient.orientation)
            << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(1));

  std::cout << "Reloading the original orientation." << std::endl;
  SetOrientation(ren, originalOrient);
  renWin->Render();
  Orientation check = GetOrientation(ren);
  std::cout << std::setw(23)
            << "Final orientation:" << CommaSeparatedList(check.orientation)
            << std::endl;
  std::this_thread::sleep_for(std::chrono::seconds(1));

  iren->Initialize();
  iren->Start();

  return EXIT_SUCCESS;
}

namespace {
std::string CommaSeparatedList(std::vector<double> const& v,
                               const int precision, const int width)
{
  std::ostringstream os;
  for (auto p = v.begin(); p < v.end() - 1; ++p)
  {
    os << std::fixed << std::setw(width) << std::setprecision(precision) << *p
       << ", ";
  }
  os << std::fixed << std::setw(width) << std::setprecision(precision)
     << v.back();
  return os.str();
}

Orientation GetOrientation(vtkRenderer* ren)
{
  Orientation p;

  vtkCamera* camera = ren->GetActiveCamera();
  p.ToVector(camera->GetPosition(), p.position);
  p.ToVector(camera->GetFocalPoint(), p.focalPoint);
  p.ToVector(camera->GetViewUp(), p.viewUp);
  p.distance = camera->GetDistance();
  p.ToVector(camera->GetClippingRange(), p.clipplingRange);
  p.ToVector(camera->GetOrientation(), p.orientation);

  return p;
}

void SetOrientation(vtkRenderer* ren, Orientation const& p)
{
  vtkCamera* camera = ren->GetActiveCamera();
  camera->SetPosition(p.position.data());
  camera->SetFocalPoint(p.focalPoint.data());
  camera->SetViewUp(p.viewUp.data());
  camera->SetDistance(p.distance);
  camera->SetClippingRange(p.clipplingRange.data());
}

} // namespace

File: src/Python/Utilities/ResetCameraOrientation.py
from time import sleep

# noinspection PyUnresolvedReferences
import vtkmodules.vtkRenderingOpenGL2
from vtkmodules.vtkCommonColor import vtkNamedColors
from vtkmodules.vtkFiltersSources import vtkConeSource
# noinspection PyUnresolvedReferences
from vtkmodules.vtkInteractionStyle import (
    vtkInteractorStyleTrackballCamera)
from vtkmodules.vtkRenderingCore import (
    vtkActor,
    vtkPolyDataMapper,
    vtkRenderWindow,
    vtkRenderWindowInteractor,
    vtkRenderer
)


def comma_separated_list(v, precision=6, width=10):
    """
    Produce a comma-separated string of numbers from a list.
    :param v: The list of floats.
    :param precision: Number of decimal places.
    :param width: The width.
    :return: A comma-separated string.
    """
    res = list()
    for p in v:
        res.append(f'{p:{width}.{precision}f}')
    return ','.join(res)


def get_orientation(ren):
    """
    Get the camera orientation.
    :param ren: The renderer.
    :return: The orientation parameters.
    """
    p = dict()
    camera = ren.GetActiveCamera()
    p['position'] = camera.GetPosition()
    p['focal point'] = camera.GetFocalPoint()
    p['view up'] = camera.GetViewUp()
    p['distance'] = camera.GetDistance()
    p['clipping range'] = camera.GetClippingRange()
    p['orientation'] = camera.GetOrientation()
    return p


def set_orientation(ren, p):
    """
    Set the orientation of the camera.
    :param ren: The renderer.
    :param p: The orientation parameters.
    :return:
    """
    camera = ren.GetActiveCamera()
    camera.SetPosition(p['position'])
    camera.SetFocalPoint(p['focal point'])
    camera.SetViewUp(p['view up'])
    camera.SetDistance(p['distance'])
    camera.SetClippingRange(p['clipping range'])


def main(argv):
    colors = vtkNamedColors()

    cone = vtkConeSource()
    cone.SetHeight(3.0)
    cone.SetRadius(1.0)
    cone.SetResolution(10)

    cone_mapper = vtkPolyDataMapper()
    cone_mapper.SetInputConnection(cone.GetOutputPort())

    cone_actor = vtkActor()
    cone_actor.SetMapper(cone_mapper)
    cone_actor.GetProperty().SetColor(colors.GetColor3d('Bisque'))

    ren = vtkRenderer()
    ren.AddActor(cone_actor)
    ren.SetBackground(colors.GetColor3d('MidnightBlue'))

    ren_win = vtkRenderWindow()
    ren_win.AddRenderer(ren)
    ren_win.SetSize(600, 600)
    ren_win.SetWindowName('ResetCameraOrientation')

    iren = vtkRenderWindowInteractor()
    iren.SetRenderWindow(ren_win)

    style = vtkInteractorStyleTrackballCamera()
    iren.SetInteractorStyle(style)

    camera = ren.GetActiveCamera()
    camera.SetRoll(15)
    camera.Elevation(-15)
    camera.Azimuth(30)
    ren.ResetCamera()

    ren_win.Render()
    original_orient = get_orientation(ren)
    s = f'{"Original orientation:":23s}'
    s += comma_separated_list(original_orient["orientation"])
    print(s)
    sleep(1)

    camera.SetPosition(-3.568189, 5.220048, 2.352639)
    camera.SetFocalPoint(-0.399044, -0.282865, 0.131438)
    camera.SetViewUp(0.623411, 0.573532, -0.531431)
    camera.SetDistance(6.727500)
    camera.SetClippingRange(3.001430, 11.434082)
    # No need to use ren.ResetCamera() as we have all the parameters.
    ren_win.Render()
    new_orient = get_orientation(ren)
    s = f'{"New orientation:":23s}'
    s += comma_separated_list(new_orient["orientation"])
    print(s)
    sleep(1)

    print('Reloading the original orientation.')
    set_orientation(ren, original_orient)
    ren_win.Render()
    check = get_orientation(ren)
    s = f'{"Final orientation:":23s}'
    s += comma_separated_list(check["orientation"])
    print(s)
    sleep(1)

    iren.Initialize()
    iren.Start()


if __name__ == '__main__':
    import sys

    main(sys.argv)
